feat(governance): Introduce Core `StrategyV1` and Token Adapter Infrastructure

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/269

Fixes [ENG-931](https://linear.app/decent-labs/issue/ENG-931/implement-foundational-strategyv1-contract-and-token-adapter)

**High-Level Context:**

This Pull Request introduces the foundational contracts for our new modular governance strategy system. This is a key part of a larger initiative to move towards a more flexible and extensible voting architecture. The core components introduced here are the `StrategyV1` contract, which will serve as the central voting engine, the `ITokenAdapterBaseV1` interface defining how different token types can plug into this engine, and an initial `MockTokenAdapter` for testing. This work prepares `AzoriusV1` (modified in a previous, related PR) to interact with this new `StrategyV1`.

**Changes in this PR:**

1.  **`StrategyV1.sol` (New Contract):**
    *   **Core Functionality:** Defines a new central voting strategy contract.
    *   **Inheritance:** Inherits from `Initializable`, `IStrategyV1`, `OwnableUpgradeable`, `UUPSUpgradeable`, `ERC165`, `ERC4337VoterSupportV1`, and `Version`.
    *   **State Variables:**
        *   `azorius`: Address of the `AzoriusV1` module it serves.
        *   `votingPeriod`: Duration for proposals.
        *   `quorumThreshold`: Fixed vote count for quorum.
        *   `basisNumerator`: For basis calculation (denominator is `BASIS_DENOMINATOR = 1_000_000`).
        *   `proposalVotingDetails`: Mapping `(uint32 => ProposalVotingDetails)` to store `votingStartTimestamp`, `votingEndTimestamp`, `votingStartBlock`, and vote counts (`yesVotes`, `noVotes`, `abstainVotes`).
        *   `tokenAdapters`: An array `ITokenAdapterBaseV1[]` to store configured token adapters.
    *   **`VoteType` Enum:** Defined for `NO`, `YES`, `ABSTAIN`.
    *   **Events:** `StrategyParametersUpdated`, `AdapterAdded`, `AdapterRemoved`, `Voted`, `ProposalInitialized`.
    *   **Errors:** Defines various errors for invalid operations and states.
    *   **Constructor:** Calls `_disableInitializers()`.
    *   **`initialize(...)`:**
        *   Sets initial owner, Azorius address, voting parameters (`_votingPeriod`, `_quorumThreshold`, `_basisNumerator`), initial token adapters, and LightAccountFactory.
        *   Validates `_azorius` is not zero.
        *   Calls `_updateVotingPeriod`, `_updateQuorumThreshold`, `_updateBasisNumerator`.
        *   Iterates through `_initialTokenAdapters` calling `_addAdapter` for each.
        *   Emits `StrategyParametersUpdated`.
    *   **`_authorizeUpgrade(...)`:** Implemented for UUPS, owner-only.
    *   **Parameter Updaters (Owner-only):** `updateVotingPeriod`, `updateQuorumThreshold`, `updateBasisNumerator`, with internal helpers `_updateVotingPeriod`, `_updateQuorumThreshold`, `_updateBasisNumerator` that include validation logic (e.g., `_newVotingPeriod > 0`, `_newBasisNumerator` within valid range).
    *   **Token Adapter Management (Owner-only):**
        *   `addAdapter(address _adapter)`: Adds an adapter to `tokenAdapters` after checking for zero address and duplicates. Emits `AdapterAdded`.
        *   `removeAdapter(address _adapter)`: Removes an adapter. Emits `AdapterRemoved`.
        *   `getTokenAdapterCount()`: Returns `tokenAdapters.length`.
    *   **`IStrategyV1` Implementation:**
        *   `initializeProposal(uint32 proposalId)`: Called by `azorius`. Sets up `proposalVotingDetails` (timestamps, start block, vote counts). Emits `ProposalInitialized`. Reverts if `msg.sender != azorius` or `tokenAdapters.length == 0`.
        *   `isPassed(uint32 _proposalId)`: Returns `true` if voting period is over and quorum & basis are met (currently, the `<main-diff>` version only checks if `block.timestamp <= proposal.votingEndTimestamp` and returns `false` if so, otherwise `true`. The quorum/basis logic seems to be in the `<context-diff>` but not this specific one). Reverts if proposal not initialized.
        *   `isProposer(address _address)`: Iterates `tokenAdapters`, calling `isProposer` on each. Returns `true` if any adapter returns `true`. Returns `false` if no adapters.
        *   `getVotingTimestamps(uint32 _proposalId)`: Returns `proposalVotingDetails[_proposalId].votingStartTimestamp` and `votingEndTimestamp`. Reverts if proposal not initialized.
    *   **Version & ERC165:** Implements `getVersion()` and `supportsInterface()`.
    *   **(Note:** The `vote` function and more detailed `isPassed` logic including `isQuorumMet` and `isBasisMet` are part of the `<context-diff>` but not this specific `<main-diff>`. The `<main-diff>` `StrategyV1.sol` is a more foundational version.)*

2.  **`ITokenAdapterBaseV1.sol` (New Interface):**
    *   Defines the interface for Token Adapters.
    *   Includes only `isProposer(address _proposer) external view returns (bool);`.
    *   **(Note:** The `weightOf` and `recordVote` methods are present in the `<context-diff>` version of this interface but not in the `<main-diff>`.)*

3.  **`MockTokenAdapter.sol` (New Contract):**
    *   Implements `ITokenAdapterBaseV1`.
    *   Contains mappings `weightsToReturn` and `proposerStatusToReturn` for test control.
    *   Includes setters `setWeight`, `setProposerStatus`, and `setShouldRevertOnRecordVote`.
    *   Implements `isProposer` based on `proposerStatusToReturn`.
    *   **(Note:** The `weightOf` and `recordVote` mock implementations are present in the `<context-diff>` but not this `<main-diff>`.)*

4.  **Test Suite `StrategyV1.test.ts` (New File):**
    *   Sets up signers, deploys `StrategyV1` implementation and proxy, `MockLightAccountFactory`, and `MockTokenAdapter` instances.
    *   Includes a `deployStrategyProxy` helper function.
    *   **Initialization Tests:**
        *   Verifies correct parameter setting and event emission.
        *   Tests reverts for invalid `azorius` address, `votingPeriod`, and `basisNumerator`.
        *   Tests initialization with zero quorum and valid min/max basis.
    *   **Adapter Management Tests:**
        *   Tests `addAdapter`: owner can add, non-owner cannot, reverts on zero address or duplicate adapter. Allows adding multiple.
        *   Tests `removeAdapter`: owner can remove, non-owner cannot, reverts on zero address or if adapter not found or list is empty. Covers removing first, last, and only adapter.
        *   Tests `getTokenAdapterCount`.
    *   **`initializeProposal` Tests:**
        *   Reverts if called by non-azorius or if no adapters are configured.
        *   Verifies correct initialization of `proposalVotingDetails` and `ProposalInitialized` event emission.
        *   Checks re-initialization resets vote counts.
    *   **`isProposer` Tests:**
        *   Returns `false` if no adapters.
        *   Returns `true` if any adapter identifies as proposer.
        *   Returns `false` if no adapter identifies as proposer.
    *   **`getVotingTimestamps` & `getProposalBlocks` Tests:** (The `<main-diff>` `StrategyV1.sol` doesn't have `getProposalBlocks`, so tests would only be for `getVotingTimestamps`).
        *   Verifies correct timestamps after initialization.
        *   Reverts if proposal not initialized.
    *   **`isPassed` Tests:** (Based on the `<main-diff>`'s simpler `isPassed` logic).
        *   Reverts if proposal not initialized.
    *   **ERC165 and Version Tests.**
    *   **UUPS Upgradeability Tests.**
    *   **Parameter Update Tests:** For `updateVotingPeriod`, `updateQuorumThreshold`, `updateBasisNumerator`.
    *   **(Note:** Tests for the full `vote` functionality, `isQuorumMet`, `isBasisMet`, and more complex `isPassed` scenarios are dependent on the `StrategyV1` version from the `<context-diff>` and would not be covered by tests based *only* on this `<main-diff>`.)*

**Impact and Status:**

*   This PR introduces the core contracts for the new modular strategy architecture (`StrategyV1`, `ITokenAdapterBaseV1`, `MockTokenAdapter`).
*   **Compilation Status:** The contracts in *this specific diff* (`StrategyV1.sol`, `ITokenAdapterBaseV1.sol`, `MockTokenAdapter.sol`) should compile. However, `StrategyV1` implements `IStrategyV1`, and if the `IStrategyV1` interface (defined in another file, presumably already updated as per `<context-diff>`) has methods not yet implemented in *this version* of `StrategyV1.sol` (e.g., the full `vote` method, `isQuorumMet`, `isBasisMet`, `getVotingStartBlock`), then `StrategyV1.sol` itself might not compile against the *latest* `IStrategyV1` interface from the `<context-diff>`. The test file `StrategyV1.test.ts` will also only be able to test the functions present in *this* `<main-diff>` version of `StrategyV1.sol`.
*   **Expected Failures:** Given this is an foundational piece, the overall project and test suite will still have significant failures until `StrategyV1` is fully implemented (as per `<context-diff>`) and all downstream dependencies and tests are updated.

adamgall commented on contracts/deployables/strategies/StrategyV1.sol:
```diff
@@ -0,0 +1,381 @@
+        proposal.yesVotes = 0;
+        proposal.noVotes = 0;
+        proposal.abstainVotes = 0;
+        // Note: mapping(address => bool) hasVoted is implicitly false for all addresses initially
```
delete this

adamgall replied:
fixed

adamgall commented on contracts/mocks/MockTokenAdapter.sol:
```diff
@@ -0,0 +1,35 @@
+
+    bool public shouldRevertRecordVote;
+
+    // --- Admin functions to set mock behavior ---
+    function setWeight(address _voter, uint256 _weight) external {
+        weightsToReturn[_voter] = _weight;
+    }
+
+    function setProposerStatus(address _proposer, bool _isProposer) external {
+        proposerStatusToReturn[_proposer] = _isProposer;
+    }
+
+    function setShouldRevertOnRecordVote(bool _shouldRevert) external {
+        // New function
+        shouldRevertRecordVote = _shouldRevert;
+    }
```
Don't need these yet in this PR.

adamgall commented on contracts/mocks/MockTokenAdapter.sol:
```diff
@@ -0,0 +1,35 @@
+        proposerStatusToReturn[_proposer] = _isProposer;
+    }
+
+    function setShouldRevertOnRecordVote(bool _shouldRevert) external {
+        // New function
+        shouldRevertRecordVote = _shouldRevert;
+    }
```
Don't need this yet in this PR.

adamgall replied:
fixed

adamgall commented on contracts/mocks/MockTokenAdapter.sol:
```diff
@@ -0,0 +1,35 @@
+    bool public shouldRevertRecordVote;
+
+    // --- Admin functions to set mock behavior ---
+    function setWeight(address _voter, uint256 _weight) external {
+        weightsToReturn[_voter] = _weight;
+    }
```
Don't need this yet in this PR.

adamgall replied:
fixed

twodam commented on contracts/deployables/strategies/StrategyV1.sol:
```diff
@@ -0,0 +1,290 @@
+        ABSTAIN
+    }
+
+    event StrategyParametersUpdated(
+        uint32 votingPeriod,
+        uint256 quorumThreshold,
+        uint256 basisNumerator
+    );
+    event AdapterAdded(address indexed adapter, uint256 index);
+    event AdapterRemoved(address indexed adapter, uint256 index);
```
should be moved to interface `IStrategyV1`?

adamgall replied:
yes agreed. for the sake of my own sanity and not getting stuck in a rebase / merge / conflict resolution hell, I'm going to do this in a new PR.

twodam commented on contracts/deployables/strategies/StrategyV1.sol:
```diff
@@ -0,0 +1,290 @@
+    );
+
+    error InvalidAzoriusAddress();
+    error InvalidVotingPeriod();
+    error InvalidBasisNumerator();
+    error AdapterIsZeroAddress();
+    error AdapterAlreadyExists();
+    error AdapterNotFound();
```
should be moved to interface `IStrategyV1`?

adamgall replied:
yes agreed. for the sake of my own sanity and not getting stuck in a rebase / merge / conflict resolution hell, I'm going to do this in a new PR.

twodam commented on contracts/deployables/strategies/StrategyV1.sol:
```diff
@@ -0,0 +1,290 @@
+    ) external override onlyOwner {
+        _updateVotingPeriod(_newVotingPeriod);
+        emit StrategyParametersUpdated(
+            votingPeriod,
```
should be `_newVotingPeriod`?

adamgall replied:
at the point when this event is emitted, `votingPeriod` and `_newVotingPeriod` are the same value.

twodam commented on contracts/deployables/strategies/StrategyV1.sol:
```diff
@@ -0,0 +1,290 @@
+        _updateQuorumThreshold(_newQuorumThreshold);
+        emit StrategyParametersUpdated(
+            votingPeriod,
+            quorumThreshold,
```
should be `_newQuorumThreshold`?

adamgall replied:
at the point when this event is emitted, `votingPeriod` and `_newVotingPeriod` are the same value.

twodam commented on contracts/deployables/strategies/StrategyV1.sol:
```diff
@@ -0,0 +1,290 @@
+        emit StrategyParametersUpdated(
+            votingPeriod,
+            quorumThreshold,
+            basisNumerator
```
should be `_newBasisNumerator`?

adamgall replied:
at the point when this event is emitted, `votingPeriod` and `_newVotingPeriod` are the same value.

twodam commented on contracts/deployables/strategies/StrategyV1.sol:
```diff
@@ -0,0 +1,290 @@
+                revert AdapterAlreadyExists();
+        }
+        tokenAdapters.push(ITokenAdapterBaseV1(_adapter));
+        emit AdapterAdded(address(_adapter), tokenAdapters.length - 1);
```
`_adapter` is already an `address`, no need to convert

adamgall replied:
yes you're right. however, i'm not going to actually make this change here, because in a later PR i actually delete the whole concept of "adding/removing" adapters, so this code is getting deleted anyway.

twodam commented on contracts/deployables/strategies/StrategyV1.sol:
```diff
@@ -0,0 +1,290 @@
+
+        for (uint256 i = 0; i < adapterCount; i++) {
+            if (address(tokenAdapters[i]) == _adapter) {
+                tokenAdapters[i] = tokenAdapters[adapterCount - 1];
+                tokenAdapters.pop();
+                emit AdapterRemoved(address(_adapter), i);
```
What's the purpose of the index in add or remove events, given that the index is likely to change?

adamgall replied:
in retrospect you're right it's not very useful. however, i'm not going to actually make this change here, because in a later PR i actually delete the whole concept of "adding/removing" adapters, so this code is getting deleted anyway.

twodam commented on test/deployables/strategies/StrategyV1.test.ts:
```diff
@@ -0,0 +1,702 @@
+      await strategy.connect(owner).addAdapter(adapter1Addr);
+      await strategy.connect(owner).addAdapter(adapter2Addr);
+
+      // To correctly test the emitted index, we need to know the exact removal logic
+      // Assuming it swaps with last and pops:
+      // If removing adapter1Addr (at index 0), adapter2Addr (at index 1) moves to 0.
+      // The event should reflect the original index of the removed item.
+      await expect(strategy.connect(owner).removeAdapter(adapter1Addr))
+        .to.emit(strategy, 'AdapterRemoved')
+        .withArgs(adapter1Addr, 0);
```
This test also suggests that including an unstable index in the event may be unnecessary — the removed address alone should be sufficient in add/remove events.

twodam commented on test/deployables/strategies/StrategyV1.test.ts:
```diff
@@ -0,0 +1,702 @@
+    });
+  });
+
+  describe('isPassed', () => {
```
Missing a test case to verify that block.timestamp <= proposal.votingEndTimestamp?

adamgall replied:
indeed. a test for this is added in the next PR though.

adamgall committed:
feat(governance): Introduce Core `StrategyV1` and Token Adapter Infrastructure

**High-Level Context:**

This Pull Request introduces the foundational contracts for our new modular governance strategy system. This is a key part of a larger initiative to move towards a more flexible and extensible voting architecture. The core components introduced here are the `StrategyV1` contract, which will serve as the central voting engine, the `ITokenAdapterBaseV1` interface defining how different token types can plug into this engine, and an initial `MockTokenAdapter` for testing. This work prepares `AzoriusV1` (modified in a previous, related PR) to interact with this new `StrategyV1`.

**Changes in this PR:**

1.  **`StrategyV1.sol` (New Contract):**
    *   **Core Functionality:** Defines a new central voting strategy contract.
    *   **Inheritance:** Inherits from `Initializable`, `IStrategyV1`, `OwnableUpgradeable`, `UUPSUpgradeable`, `ERC165`, `ERC4337VoterSupportV1`, and `Version`.
    *   **State Variables:**
        *   `azorius`: Address of the `AzoriusV1` module it serves.
        *   `votingPeriod`: Duration for proposals.
        *   `quorumThreshold`: Fixed vote count for quorum.
        *   `basisNumerator`: For basis calculation (denominator is `BASIS_DENOMINATOR = 1_000_000`).
        *   `proposalVotingDetails`: Mapping `(uint32 => ProposalVotingDetails)` to store `votingStartTimestamp`, `votingEndTimestamp`, `votingStartBlock`, and vote counts (`yesVotes`, `noVotes`, `abstainVotes`).
        *   `tokenAdapters`: An array `ITokenAdapterBaseV1[]` to store configured token adapters.
    *   **`VoteType` Enum:** Defined for `NO`, `YES`, `ABSTAIN`.
    *   **Events:** `StrategyParametersUpdated`, `AdapterAdded`, `AdapterRemoved`, `Voted`, `ProposalInitialized`.
    *   **Errors:** Defines various errors for invalid operations and states.
    *   **Constructor:** Calls `_disableInitializers()`.
    *   **`initialize(...)`:**
        *   Sets initial owner, Azorius address, voting parameters (`_votingPeriod`, `_quorumThreshold`, `_basisNumerator`), initial token adapters, and LightAccountFactory.
        *   Validates `_azorius` is not zero.
        *   Calls `_updateVotingPeriod`, `_updateQuorumThreshold`, `_updateBasisNumerator`.
        *   Iterates through `_initialTokenAdapters` calling `_addAdapter` for each.
        *   Emits `StrategyParametersUpdated`.
    *   **`_authorizeUpgrade(...)`:** Implemented for UUPS, owner-only.
    *   **Parameter Updaters (Owner-only):** `updateVotingPeriod`, `updateQuorumThreshold`, `updateBasisNumerator`, with internal helpers `_updateVotingPeriod`, `_updateQuorumThreshold`, `_updateBasisNumerator` that include validation logic (e.g., `_newVotingPeriod > 0`, `_newBasisNumerator` within valid range).
    *   **Token Adapter Management (Owner-only):**
        *   `addAdapter(address _adapter)`: Adds an adapter to `tokenAdapters` after checking for zero address and duplicates. Emits `AdapterAdded`.
        *   `removeAdapter(address _adapter)`: Removes an adapter. Emits `AdapterRemoved`.
        *   `getTokenAdapterCount()`: Returns `tokenAdapters.length`.
    *   **`IStrategyV1` Implementation:**
        *   `initializeProposal(uint32 proposalId)`: Called by `azorius`. Sets up `proposalVotingDetails` (timestamps, start block, vote counts). Emits `ProposalInitialized`. Reverts if `msg.sender != azorius` or `tokenAdapters.length == 0`.
        *   `isPassed(uint32 _proposalId)`: Returns `true` if voting period is over and quorum & basis are met (currently, the `<main-diff>` version only checks if `block.timestamp <= proposal.votingEndTimestamp` and returns `false` if so, otherwise `true`. The quorum/basis logic seems to be in the `<context-diff>` but not this specific one). Reverts if proposal not initialized.
        *   `isProposer(address _address)`: Iterates `tokenAdapters`, calling `isProposer` on each. Returns `true` if any adapter returns `true`. Returns `false` if no adapters.
        *   `getVotingTimestamps(uint32 _proposalId)`: Returns `proposalVotingDetails[_proposalId].votingStartTimestamp` and `votingEndTimestamp`. Reverts if proposal not initialized.
    *   **Version & ERC165:** Implements `getVersion()` and `supportsInterface()`.
    *   **(Note:** The `vote` function and more detailed `isPassed` logic including `isQuorumMet` and `isBasisMet` are part of the `<context-diff>` but not this specific `<main-diff>`. The `<main-diff>` `StrategyV1.sol` is a more foundational version.)*

2.  **`ITokenAdapterBaseV1.sol` (New Interface):**
    *   Defines the interface for Token Adapters.
    *   Includes only `isProposer(address _proposer) external view returns (bool);`.
    *   **(Note:** The `weightOf` and `recordVote` methods are present in the `<context-diff>` version of this interface but not in the `<main-diff>`.)*

3.  **`MockTokenAdapter.sol` (New Contract):**
    *   Implements `ITokenAdapterBaseV1`.
    *   Contains mappings `weightsToReturn` and `proposerStatusToReturn` for test control.
    *   Includes setters `setWeight`, `setProposerStatus`, and `setShouldRevertOnRecordVote`.
    *   Implements `isProposer` based on `proposerStatusToReturn`.
    *   **(Note:** The `weightOf` and `recordVote` mock implementations are present in the `<context-diff>` but not this `<main-diff>`.)*

4.  **Test Suite `StrategyV1.test.ts` (New File):**
    *   Sets up signers, deploys `StrategyV1` implementation and proxy, `MockLightAccountFactory`, and `MockTokenAdapter` instances.
    *   Includes a `deployStrategyProxy` helper function.
    *   **Initialization Tests:**
        *   Verifies correct parameter setting and event emission.
        *   Tests reverts for invalid `azorius` address, `votingPeriod`, and `basisNumerator`.
        *   Tests initialization with zero quorum and valid min/max basis.
    *   **Adapter Management Tests:**
        *   Tests `addAdapter`: owner can add, non-owner cannot, reverts on zero address or duplicate adapter. Allows adding multiple.
        *   Tests `removeAdapter`: owner can remove, non-owner cannot, reverts on zero address or if adapter not found or list is empty. Covers removing first, last, and only adapter.
        *   Tests `getTokenAdapterCount`.
    *   **`initializeProposal` Tests:**
        *   Reverts if called by non-azorius or if no adapters are configured.
        *   Verifies correct initialization of `proposalVotingDetails` and `ProposalInitialized` event emission.
        *   Checks re-initialization resets vote counts.
    *   **`isProposer` Tests:**
        *   Returns `false` if no adapters.
        *   Returns `true` if any adapter identifies as proposer.
        *   Returns `false` if no adapter identifies as proposer.
    *   **`getVotingTimestamps` & `getProposalBlocks` Tests:** (The `<main-diff>` `StrategyV1.sol` doesn't have `getProposalBlocks`, so tests would only be for `getVotingTimestamps`).
        *   Verifies correct timestamps after initialization.
        *   Reverts if proposal not initialized.
    *   **`isPassed` Tests:** (Based on the `<main-diff>`'s simpler `isPassed` logic).
        *   Reverts if proposal not initialized.
    *   **ERC165 and Version Tests.**
    *   **UUPS Upgradeability Tests.**
    *   **Parameter Update Tests:** For `updateVotingPeriod`, `updateQuorumThreshold`, `updateBasisNumerator`.
    *   **(Note:** Tests for the full `vote` functionality, `isQuorumMet`, `isBasisMet`, and more complex `isPassed` scenarios are dependent on the `StrategyV1` version from the `<context-diff>` and would not be covered by tests based *only* on this `<main-diff>`.)*

**Impact and Status:**

*   This PR introduces the core contracts for the new modular strategy architecture (`StrategyV1`, `ITokenAdapterBaseV1`, `MockTokenAdapter`).
*   **Compilation Status:** The contracts in *this specific diff* (`StrategyV1.sol`, `ITokenAdapterBaseV1.sol`, `MockTokenAdapter.sol`) should compile. However, `StrategyV1` implements `IStrategyV1`, and if the `IStrategyV1` interface (defined in another file, presumably already updated as per `<context-diff>`) has methods not yet implemented in *this version* of `StrategyV1.sol` (e.g., the full `vote` method, `isQuorumMet`, `isBasisMet`, `getVotingStartBlock`), then `StrategyV1.sol` itself might not compile against the *latest* `IStrategyV1` interface from the `<context-diff>`. The test file `StrategyV1.test.ts` will also only be able to test the functions present in *this* `<main-diff>` version of `StrategyV1.sol`.
*   **Expected Failures:** Given this is an foundational piece, the overall project and test suite will still have significant failures until `StrategyV1` is fully implemented (as per `<context-diff>`) and all downstream dependencies and tests are updated.
